README.md
## cnfapi-miniprogram
----------------------------
author: bugszhou | Email:bugszhou@outlook.com <br>
description: A lib project with ziu

cnfapi-miniprogram是基于promise的小程序http库，可以用在微信小程序，支付宝小程序和百度小程序

### Features

* 支持 promise API
* 拦截请求和响应
* 自动转换 JSON 数据

### 安装

```shell
npm install -S cnfapi-miniprogram
```

### Example

```javascript
import Api from 'cnfapi-miniprogram';

const api = new Api({
  baseURL: 'https://prj1.demo.com',
  env: 'weapp', // 使用环境：weapp - 微信小程序、aliapp - 支付宝小程序、swan - 百度小程序
  timeout: 10000, // 10s超时
  headers: {
    'Content-Type': 'application/json',
  },
  resSuccessCallback(data, next) {
    // next接受3个参数
    // 第一个参数是代表error
    // 第二个参数是代表传递给 resolve 的数据
    // 第三个参数是自定义数据
    if (data.code === 200) {
      next(null, data.data, data.code);
    } else {
      next({
        msg: data.msg,
        retcode: data.code,
      }, {}, data.code);
    }
  },
  // 是否开启response拦截器
  openResInterceptor(res) {
    return true;
  },
  // response拦截器处理函数
  // openResInterceptor返回true执行方法
  resInterceptor(next) {
    next();
  },
}, {
     getList: {
       interval: 2000, // 每隔2秒重试一次
       retryTimes: 10, // 重试10次
       apiName: '/test/prj/getList', // 接口pathurl
       desc: '', // 接口描述
       method: 'POST',
       params: {
         // post参数
         post: [{
            param: 'param1', // 参数名
            isNeed: 1, // 是否必须 1 为必须、0为非必须
         }],
         // get参数
         get: [{
           param: 'param2', // 参数名
           isNeed: 0, // 是否必须 1 为必须、0为非必须
        }],
       },
     },
     // restful模式
     getOrders: {
       interval: 2000, // 每隔2秒重试一次
       retryTimes: 10, // 重试10次
       apiName: '/test/prj/getOrders/{orderid}', // 接口pathurl
       desc: '', // 接口描述
       method: 'POST',
       params: {},
     },
   });

   // http请求调用方式
   // 1. 常规方式
   api.getList({
     getRequestTask: (tesk) => {}, // 获取请求任务对象task，task对象可对中断请求任务等功能
     data: {
       param1: 'param1',
       param2: 'param2',
     }
   })
   .then(({ data }) => {
     // 服务端数据
     console.log(data);
   })
   .catch(err => {
     console.log(err);
   });
   // 2. restful方式
   api.getOrders({
     restful: {
       orderid: 'param1',
     }
   })
   .then(({ data }) => {
     // 服务端数据
     console.log(data);
   })
   .catch(err => {
     console.log(err);
   });

```

- task功能[点击查看](https://developers.weixin.qq.com/miniprogram/dev/api/network/request/RequestTask.html)

### request拦截器

```javascript
api._before = function _before(apiOpts, apiConf, next) {
  // apiOpts为调用时传入的参数，如：
  /*
  {
    data: {
      param1: 123
    }
  }
  */
  // apiConf 该接口调用的配置参数
  // 可在_before内对参数做处理，例如：对参数做签名保护
  next(apiOpts);
};
```

### 钩子函数

1. 发送请求前`cnfapi:req:before`

```javascript
api.on('cnfapi:req:before', (data) => {
    // 业务逻辑
});
```

2. 请求成功`cnfapi:res:resolve`

```javascript
api.on('cnfapi:res:resolve', (data) => {
    // 业务逻辑
});
```

3. 请求reject`cnfapi:res:reject`

```javascript
api.on('cnfapi:res:reject', (data) => {
    // 业务逻辑
});
```

4. 请求发生错误`cnfapi:res:catch`

```javascript
api.on('cnfapi:res:catch', (data) => {
    // 业务逻辑
});
```



CHANGELOG.md
<a name="2.1.2"></a>
## [2.1.2](https://github.com/bugszhou/cnfapi-miniprogram/compare/v2.1.1...v2.1.2) (2020-06-03)


### Features

* 成功回调增加msg ([398122f](https://github.com/bugszhou/cnfapi-miniprogram/commit/398122f))



<a name="2.1.1"></a>
## [2.1.1](https://github.com/bugszhou/cnfapi-miniprogram/compare/v2.1.0...v2.1.1) (2020-06-03)


### Features

* 修改this ([5e9f402](https://github.com/bugszhou/cnfapi-miniprogram/commit/5e9f402))



<a name="2.1.0"></a>
# [2.1.0](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.3.1...v2.1.0) (2020-05-18)


### Features

* 增加customerHttp ([1cefc0a](https://github.com/bugszhou/cnfapi-miniprogram/commit/1cefc0a))
* 钩子函数增加返回数据 ([5404918](https://github.com/bugszhou/cnfapi-miniprogram/commit/5404918))



<a name="1.3.1"></a>
## [1.3.1](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.3.0...v1.3.1) (2019-09-12)



<a name="1.3.0"></a>
# [1.3.0](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.2.4...v1.3.0) (2019-09-12)



<a name="1.2.4"></a>
## [1.2.4](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.2.3...v1.2.4) (2019-08-08)



<a name="1.2.3"></a>
## [1.2.3](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.2.2...v1.2.3) (2019-07-31)



<a name="1.2.2"></a>
## [1.2.2](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.2.1...v1.2.2) (2019-07-31)


### Bug Fixes

* Modify event name ([aacc6fc](https://github.com/bugszhou/cnfapi-miniprogram/commit/aacc6fc))



<a name="1.2.1"></a>
## [1.2.1](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.2.0...v1.2.1) (2019-07-31)


### Bug Fixes

* Add fnName ([4a92e18](https://github.com/bugszhou/cnfapi-miniprogram/commit/4a92e18))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.1.4...v1.2.0) (2019-07-31)


### Features

* Add events ([c1e0f4b](https://github.com/bugszhou/cnfapi-miniprogram/commit/c1e0f4b))



<a name="1.1.4"></a>
## [1.1.4](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.1.3...v1.1.4) (2019-07-31)


### Features

* Rm abort retry ([446059c](https://github.com/bugszhou/cnfapi-miniprogram/commit/446059c))



<a name="1.1.3"></a>
## [1.1.3](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.1.2...v1.1.3) (2019-07-31)


### Bug Fixes

* **http:** 传给requestCallBack参数不正确 ([fdc06f9](https://github.com/bugszhou/cnfapi-miniprogram/commit/fdc06f9))



<a name="1.1.2"></a>
## [1.1.2](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.1.0...v1.1.2) (2019-07-31)



<a name="1.1.0"></a>
# [1.1.0](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.0.9...v1.1.0) (2019-07-31)


### Features

* Add getRequeatTask ([ec9f596](https://github.com/bugszhou/cnfapi-miniprogram/commit/ec9f596))



<a name="1.0.9"></a>
## [1.0.9](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.0.8...v1.0.9) (2019-07-26)



<a name="1.0.8"></a>
## [1.0.8](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.0.7...v1.0.8) (2019-07-26)


### Features

* 新增before函数错误中止 ([578bbc9](https://github.com/bugszhou/cnfapi-miniprogram/commit/578bbc9))



<a name="1.0.7"></a>
## [1.0.7](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.0.6...v1.0.7) (2019-07-24)


### Bug Fixes

* Modify retryTimes ([c4c4667](https://github.com/bugszhou/cnfapi-miniprogram/commit/c4c4667))



<a name="1.0.6"></a>
## [1.0.6](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.0.5...v1.0.6) (2019-07-24)


### Bug Fixes

* Add resInterceptor error ([8535328](https://github.com/bugszhou/cnfapi-miniprogram/commit/8535328))



<a name="1.0.5"></a>
## [1.0.5](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.0.4...v1.0.5) (2019-07-24)


### Features

* 增加拦截器使用次数 ([1b56c9e](https://github.com/bugszhou/cnfapi-miniprogram/commit/1b56c9e))



<a name="1.0.4"></a>
## [1.0.4](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.0.3...v1.0.4) (2019-07-11)



<a name="1.0.3"></a>
## [1.0.3](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.0.2...v1.0.3) (2019-07-11)



<a name="1.0.2"></a>
## [1.0.2](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.0.1...v1.0.2) (2019-07-11)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/bugszhou/cnfapi-miniprogram/compare/db7278f...v1.0.1) (2019-07-11)


### Features

* Finish ([db7278f](https://github.com/bugszhou/cnfapi-miniprogram/commit/db7278f))



<a name="2.1.1"></a>
## [2.1.1](https://github.com/bugszhou/cnfapi-miniprogram/compare/v2.1.0...v2.1.1) (2020-06-03)


### Features

* 修改this ([5e9f402](https://github.com/bugszhou/cnfapi-miniprogram/commit/5e9f402))



<a name="2.1.0"></a>
# [2.1.0](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.3.1...v2.1.0) (2020-05-18)


### Features

* 增加customerHttp ([1cefc0a](https://github.com/bugszhou/cnfapi-miniprogram/commit/1cefc0a))
* 钩子函数增加返回数据 ([5404918](https://github.com/bugszhou/cnfapi-miniprogram/commit/5404918))



<a name="1.3.1"></a>
## [1.3.1](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.3.0...v1.3.1) (2019-09-12)



<a name="1.3.0"></a>
# [1.3.0](https://github.com/bugszhou/cnfapi-miniprogram/compare/v1.2.4...v1.3.0) (2019-09-12)



<a name="1.2.4"></a>

src/lib/Api.js

import merge from 'lodash/merge';
import { generator } from './generator.js';
import EventEmitter from 'events';

/**
 * 获取数据类型
 */
function getType(val) {
  return Object.prototype.toString.call(val).slice(8, -1).toLowerCase();
}

class Api extends EventEmitter{
  constructor(conf = {}, apiList = {}) {
    super();
    const listType = this.getType(apiList);
    if (this.getType(conf) !== 'object' || (listType !== 'object' && listType !== 'array')) {
      throw new Error('constructor params require Object type');
    }
    const _tempApiList = JSON.parse(JSON.stringify(apiList));
    this.merge = merge;
    this.outConf = conf;
    Object.defineProperty(this, 'apiList', {
      get() {
        return _tempApiList;
      },
    });
    this.init();
  }

  init() {
    this.mergeConf(this.initConf, this.outConf);
    generator.apply(this, [JSON.parse(JSON.stringify(this.apiList)), this.defaultOpts]);
  }

  get conf() {
    return this.mergeConf(this.initConf, this.outConf);
  }

  get initConf() {
    return {
      baseURL: {
        required: true,
        errMsg: 'baseURL is required',
        vaildFn(val) {
          return {

            result: /^((http:\/\/)|(https:\/\/)|(:\/\/))([a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}(\/)?/.test(val) || /^((http:\/\/)|(https:\/\/)|(:\/\/))(localhost)/.test(val),
            errMsg: 'invalid baseURL',
          };
        },
      },
      timeout: {
        required: true,
        errMsg: 'timeout is required or not 0',
        default() {
          return 3000;
        },
        vaildFn(val) {
          return {
            result: !isNaN(val) && val != 0,
            errMsg: 'invalid timeout',
          };
        },
      },
      env: {
        required: true,
        errMsg: 'invalid env',
        default() {
          return 'browser';
        },
        vaildFn(val) {
          let list = ['browser', 'aliapp', 'weapp', 'swan'],
            resultObj = {
              result: list.indexOf(val) > -1,
              errMsg: 'invalid env, env must in ["browser", "aliapp" ,"weapp", "swan"]',
            };
          if (val === 'aliapp' || val === 'weapp') {
            window = undefined;
            document = undefined;
          }
          if (val === 'browser') {
            resultObj = {
              result: window && document && typeof window !== 'undefined' && typeof document !== 'undefined',
              errMsg: 'invalid env, env not in browser, env must in ["browser", "aliapp" ,"weapp", "swan"]',
            };
          }
          return resultObj;
        },
      },
      openResInterceptor: {
        required: false,
        default() {
          return function() {
            return false;
          };
        },
      },
      resInterceptor: {
        required: false,
        default() {
          return function() {};
        },
      },
      resSuccessCallback: {
        required: false,
        default() {
          return function(serverData, next) {
            next(false, serverData);
          };
        },
      },
      resFormat: {
        description: '接口返回格式',
        required: false,
        default() {
          return {
            type: 'object',
            properties: {
              retcode: {
                type: 'string',
              },
              msg: {
                type: 'string',
              },
              data: {
                type: 'object',
              },
            },
          };
        },
      },
    };
  }

  get defaultOpts() {
    const opts = {
      method: 'GET',
      headers: {
        // 'X-Requested-With': 'XMLHttpRequest',
        'Content-Type': 'application/x-www-form-urlencoded',
      },
    };
    return this.merge(opts, this.outConf, this.conf);
  }

  /**
     * [mergeConf 合并默认参数]
     * @author youzhao.zhou
     * @date   2018-01-24T19:21:59+0800
     * @param  {[type]}                 obj     [元对象]
     * @param  {[type]}                 sources [需要合并对象]
     * @return {Object}                         [返回合并对象]
     */
  mergeConf(obj, sources) {
    const temp = {};
    if (this.getType(obj) !== 'object' || this.getType(sources) !== 'object') {
      throw new Error('mergeConf argument Not Object');
    }
    Object.keys(obj).forEach((key, i) => {
      let val = obj[key],
        vaildResult = null,
        { vaildFn } = obj[key],
        defaultVal = (typeof obj[key].default === 'function' && obj[key].default.apply(this)) || '';
      if (val.required && !defaultVal && !sources[key]) {
        throw new Error(`${val.errMsg}`);
      }
      temp[key] = sources[key] || defaultVal;
      if (typeof vaildFn === 'function') {
        vaildResult = vaildFn(temp[key]);
      }
      if (vaildResult != null && !vaildResult.result) {
        throw new Error(`${vaildResult.errMsg}`);
      }
    });
    return temp;
  }

  openResInterceptor() {
    return this.conf.openResInterceptor.apply(this, arguments);
  }

  /**
     * 请求前统一调用
     * @param apiConf
     * @param cb
     * @private
     */
  _before(apiConf, cb) {
    cb(apiConf);
  }

  /**
     * 请求成功回调函数
     */
  // resSuccessCallback() {}

  /**
     * [getType 获取数据类型]
     * @author youzhao.zhou
     * @date   2018-01-25T11:46:48+0800
     * @param  {Any}                 val [需要获取类型的值]
     * @return {String}                     [val类型]
     */
  getType(val) {
    return getType(val);
  }
}
// import apiList from '../api/apiLists.js';
// let test = new Api({
//     baseURL: 'https://prj-test.xxx.com',
//     timeout: 0,
//     openResInterceptor(res) {
//         if (res.retcode === 5000) {
//             return true;
//         }
//         return false;
//     }
// }, apiList);
// test.initApi({
//     data: {
//         ticket: '5555',
//         openid: '345345345345'
//     },
//     resInterceptor(res) {
//         console.log(res, ' ----> ');
//     }
// }).then(() => {}, (err) => {
//     console.log(err);
// }).catch((cat) => {
//     console.log(cat, ' ------> cat');
// });
// console.log(test.conf.openResInterceptor);
export default Api;


rollup.config.js
import path from 'path';
import resolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import babel from 'rollup-plugin-babel';
import { terser } from "rollup-plugin-terser";

export default {
  external: ['lodash', 'url-parse', 'events', 'promise'],
  input: 'src/entry/index.js',
  output: [
    {
      name: 'cnfapi-miniprogram',
      file: 'dist/cnfapi-miniprogram.js',
      format: 'umd',
      sourcemap: true,
      strict: true,
      noConflict: true,
    },
    {
      name: 'cnfapi-miniprogram',
      file: 'dist/cnfapi-miniprogram.common.js',
      format: 'umd',
      sourcemap: true,
      strict: true,
      noConflict: true,
    },
    {
      file: 'dist/cnfapi-miniprogram.esm.js',
      format: 'esm',
      sourcemap: true,
      strict: true,
    },
  ],
  plugins: [
    resolve(),
    commonjs(),
    babel({
      configFile: path.resolve(__dirname, './src/entry/.babelrc'),
      runtimeHelpers: true,
      exclude: 'node_modules/**',
    }),
    terser({
      include: [/^.+\.common\.js$/],
    }),
  ],
};


src/lib/generator.js
/**
 * 循环遍历对象
 * 以对象属性名为函数名
 * 获取配置
 * 更改本次baseURL
 * 效验参数
 * 拼接get参数
 * 拼接post参数
 */


import Promise from 'promise';
import urlParse from 'url-parse';
import merge from 'lodash/merge';
import assign from 'lodash/assign';
import jsonSchema2api from './jsonSchema2api';
import Proxy from './Proxy';
import http from './http';

/**
 * errCode
 */

const retcode = {
  OK: 'FE-200',
  PARAM: 'FE-5000',
  OTHER: 'FE-5001',
  CATCH: 'FE-5002',
};

export function generator(list = {}, opts = {}, _this = {}) {
  let listKeys = list,
    isArray = Array.isArray(list);
  if (!isArray) {
    listKeys = Object.keys(list);
  }
  _this = this || _this;

  listKeys.forEach((item) => {
    let fnName = isArray ? item.name : item,
      listVal = isArray ? item : list[item];

    listVal.name = fnName;
    if (!fnName) {
      throw new Error(`Function name is required!`);
    }
    if (_this[fnName]) {
      throw new Error(`Duplicate statements in _this: ${fnName}`);
    }

    _this[fnName] = (function(conf) {
      const fn = function(apiOpts = {}, apiConfig, cb = () => {
      }) {
        apiOpts.fnName = fnName;
        apiOpts.openResInterceptor = apiOpts.openResInterceptor || opts.openResInterceptor;
        apiOpts.resInterceptor = apiOpts.resInterceptor || opts.resInterceptor;
        apiOpts.resSuccessCallback = apiOpts.resSuccessCallback || opts.resSuccessCallback;
        apiOpts.customerHttp = apiOpts.customerHttp || opts.customerHttp;
        _this._before = _this._before || opts._before;
        apiConfig.url = apiOpts.url || apiConfig.url;

        if (typeof _this._before === 'function') {
          return _this._before(apiOpts, apiConfig, (moreData) => {
            apiOpts = merge(apiOpts, moreData);
            serialize(apiOpts.data || {}, apiConfig.params, (retData) => {
              cb(retData, apiOpts);
            });
          });
        }

        serialize(apiOpts.data || {}, apiConfig.params, (retData) => {
          cb(retData, apiOpts);
        });
      };
      return getProxy(fn, JSON.parse(JSON.stringify(getConfig(conf, opts, fn))));
    }(listVal));

    listVal = null;
  });

  return _this;
}

/**
 * [hasBaseURL 接口配置参数中是否有baseURL]
 * @author youzhao.zhou
 * @date   2018-02-06T17:55:55+0800
 * @param  {String}                 url [接口地址]
 * @return {Boolean}                    [true -- 包含， false -- 不包含]
 */
function hasBaseURL(url) {
  return /^((http:\/\/)|(https:\/\/)|(:\/\/))/.test(url) || /^(localhost)/.test(url);
}

/**
 * [getConfig 接口配置和默认配置合并]
 * @author youzhao.zhou
 * @date   2018-02-06T17:57:03+0800
 * @param  {Object}                 conf        [单个接口的配置]
 * @param  {Object}                 defaultConf [默认配置]
 * @return {Object}                             [合并后的配置]
 */
function getConfig(conf, defaultConf, proxy) {
  let apiConfig = {
      url: '',
      baseURL: defaultConf.baseURL,
      env: defaultConf.env,
      headers: defaultConf.headers,
      timeout: defaultConf.timeout,
      method: conf.method || defaultConf.method,
      model: conf.resSchema || {},
      interval: conf.interval || 0, // 重试间隔时间
      retryTimes: conf.retryTimes || 0, // 重试次数
      pathname: getPathname(conf.apiName),
      params: getParams(conf.params),
      fnName: conf.name,
      signKey: conf.signKey,
      status: '',
      statusText: '',
    },
    apiName = urlParse(conf.apiName);
  apiConfig.url = apiConfig.baseURL + apiConfig.pathname;

  if (hasBaseURL(conf.apiName)) {
    apiConfig = assign(apiConfig, {
      baseURL: apiName.origin,
      pathname: getPathname(apiName.pathname),
      url: apiName.href,
    });
  }
  Object.keys(apiConfig).forEach((key) => {
    if (proxy[key]) {
      throw new Error(`Duplicate statements in proxy Function: ${key}`);
    }
    proxy[key] = apiConfig[key];
  });
  return merge({}, conf, apiConfig);
}

function getRestfulUrl(url = '', data = {}) {
  const re = /\{(.+?)\}/g;
  let result = null;
  do {
    result = re.exec(url);
    if (result && result.length > 1) {
      url = url.replace(result[0], data[result[1]] || '');
    }
  } while (result);
  return url;
}

/**
 * [getParams 将接口配置文件(apiList)中的params重构]
 * 数据结构：
 * {
 *     apiFnName: {
 *         required: true/false,
 *         method: 'POST'/'GET'
 *     }
 * }
 * @author youzhao.zhou
 * @date   2018-02-06T17:58:38+0800
 * @param  {Object}                 params [参数配置对象]
 * @return {Object}                        [重构后的对象]
 */
function getParams(params) {
  const temp = {},
    getParams = params.get || params.GET || [],
    postParams = params.post || params.POST || [];

  getParams.forEach((item) => {
    temp[item.param] = setData(item, 'GET');
  });
  postParams.forEach((item) => {
    temp[item.param] = setData(item, 'POST');
  });
  return temp;

  function setData(item, method) {
    return {
      required: !!item.isNeed,
      method: method.toUpperCase(),
    };
  }
}

/**
 * [serialize 序列化传入后端参数]
 * @author youzhao.zhou
 * @date   2018-02-06T18:01:30+0800
 * @param  {Object}                 data  [入参数据]
 * @param  {Object}                 vaild [有效验证对象]
 * @param  {Function}               cb    [验证结束回调函数]
 */
function serialize(data, vaild, cb = () => {
}) {
  setTimeout(() => {
    const retData = {
        retcode: 'FE-5000',
        errMsg: '',
        data: {},
      },
      qs = [],
      getData = {},
      postData = {},
      isVaild = Object.keys(vaild).every((param) => {
        const item = vaild[param],
          { method } = item,
          { required } = item,
          val = data[param],
          vaildResult = isVaildFn(val, param, required);
        if (!vaildResult.result) {
          retData.errMsg = vaildResult.errMsg;
          return false;
        }

        if (typeof val === 'undefined' || val == null) {
          return true;
        }

        if (method.toUpperCase() === 'GET') {
          getData[param] = val;
          qs.push(`${param}=${encodeURIComponent(val)}`);
        } else {
          postData[param] = val;
        }

        return true;
      });
    if (isVaild) {
      retData.retcode = retcode.OK;
      retData.data = {
        qs: qs.join('&'),
        postData,
        getData,
      };
    } else {
      retData.retcode = retcode.PARAM;
    }
    cb(retData);
  }, 0);
}

function isVaildFn(val, key, required) {
  if (required && (val === null || typeof val === 'undefined')) {
    return {
      result: false,
      errMsg: `param: ${key}. Is Required!`,
    };
  }
  return {
    result: true,
    errMsg: '',
  };
}

/**
 * [getPathname 格式化pathname]
 * @author youzhao.zhou
 * @date   2018-02-06T17:51:58+0800
 * @param  {String}                 pathname [路径]
 * @return {String}                          [格式化后的路径]
 */
function getPathname(pathname) {
  return (/^\//).test(pathname) ? pathname : `/${pathname}`;
}

/**
 * [getProxy 请求接口代理对象]
 * @author youzhao.zhou
 * @date   2018-02-06T17:52:41+0800
 * @param  {Function}               fn        [需要代理的接口函数]
 * @param  {Object}                 apiConfig [该接口的初始化配置参数]
 * @return {Proxy}
 */
function getProxy(fn, apiConfig = {}) {
  return new Proxy(fn, {
    get(target, name) {
      return apiConfig[name];
    },
    set() {
      throw new Error('The property is readonly!');
    },
    apply(target, ctx, args) {
      return new Promise(((resolve, reject) => {
        let { interval } = apiConfig,
          { retryTimes } = apiConfig,
          reqTime = 0;
        request(args[0] || {});

        function request(reqData) {
          ++reqTime;
          target(reqData || {}, apiConfig, (retData, apiOpts) => {
            /**
             * [retData 序列化后的入参对象]
             * retData.qs 序列化后的GET参数
             * retData.getData  get参数
             * retData.postData  post参数
             */
            /**
             * [apiOpts 入参对象]
             */
            const { data } = retData;

            apiConfig = merge(apiConfig, {
              headers: apiOpts.headers,
              timeout: apiOpts.timeout,
              fnName: apiOpts.fnName,
            });

            if (apiOpts.err) {
              ctx.emit('cnfapi:res:reject', {
                fnName: apiOpts.fnName,
                retcode: apiOpts.err.retcode,
                msg: apiOpts.err.msg,
                headers: apiConfig.headers,
                req: {
                  headers: apiConfig.headers,
                  data: data.postData,
                  qs: data.qs,
                  params: data.getData,
                },
              });
              return reject({
                retcode: apiOpts.err.retcode,
                msg: apiOpts.err.msg,
                headers: apiConfig.headers,
              });
            }

            if (retData.retcode !== retcode.OK) {
              ctx.emit('cnfapi:res:reject', {
                fnName: apiOpts.fnName,
                retcode: retData.retcode,
                msg: retData.errMsg,
                headers: apiConfig.headers,
                req: {
                  headers: apiConfig.headers,
                  data: data.postData,
                  qs: data.qs,
                  params: data.getData,
                },
              });
              return reject({
                retcode: retData.retcode,
                msg: retData.errMsg,
                headers: apiConfig.headers,
              });
            }

            let reqUrl = apiConfig.url;
            if (apiOpts.restful) {
              reqUrl = getRestfulUrl(reqUrl, apiOpts.restful);
            }
            ctx.emit('cnfapi:req:before', {
              fnName: apiOpts.fnName,
              url: reqUrl,
              timeout: apiConfig.timeout,
              env: apiConfig.env,
              method: apiConfig.method,
              headers: apiConfig.headers,
              data: data.postData,
              qs: data.qs,
              params: data.getData,
              getRequestTask: apiOpts.getRequestTask,
            });
            let reqHttp = http;
            if (apiOpts.customerHttp && typeof apiOpts.customerHttp === 'function') {
              reqHttp = apiOpts.customerHttp;
            }
            reqHttp({
              url: reqUrl,
              timeout: apiConfig.timeout,
              env: apiConfig.env,
              method: apiConfig.method,
              headers: apiConfig.headers,
              data: data.postData,
              qs: data.qs,
              params: data.getData,
              getRequestTask: apiOpts.getRequestTask,
            }).then((res) => {
              /**
               * [res 数据结构]
               * {
                             *     data: {},
                             *     headers: {},
                             *     status: 200,
                             *     statusText: 'ok'
                             * }
               */
              const serverData = res.data;
              apiConfig.status = res.status;
              apiConfig.statusText = res.statusText;

              const isOpenResInterceptor = typeof apiOpts.openResInterceptor === 'function' && apiOpts.openResInterceptor.call(apiConfig, serverData);
              if (isOpenResInterceptor && reqTime < (retryTimes + (retryTimes ? 1 : 2))) {
                return apiOpts.resInterceptor.call(apiConfig, serverData, (err, nOpts = {}) => {
                  if (err) {
                    ctx.emit('cnfapi:res:reject', {
                      fnName: apiOpts.fnName,
                      req: {
                        headers: apiConfig.headers,
                        data: data.postData,
                        qs: data.qs,
                        params: data.getData,
                      },
                      ...err
                    });
                    return reject(err);
                  }
                  const data = merge(reqData.data, nOpts.data),
                    headers = merge(apiOpts.headers, nOpts.headers);
                  reqData.data = data;
                  apiOpts.headers = headers;

                  request(reqData);
                });
              }
              if (typeof apiOpts.resSuccessCallback === 'function') {
                return apiOpts.resSuccessCallback.bind(apiConfig)(serverData, (err, resData, retcode = 200) => {
                  if (!err) {
                    reqTime = 0;
                    /**
                     * [如果该接口有model配置，则根据model转换后输出]
                     */
                    if (!isEmpty(apiConfig.model)) {
                      resData = modelFn(apiConfig.model, resData);
                    }
                    ctx.emit('cnfapi:res:resolve', {
                      fnName: apiOpts.fnName,
                      req: {
                        headers: apiConfig.headers,
                        data: data.postData,
                        qs: data.qs,
                        params: data.getData,
                      },
                      ...success({
                        data: resData,
                        headers: res.headers,
                        retcode,
                      }),
                    });
                    const msgData = {};
                    if (serverData.msg) {
                      msgData.msg = serverData.msg;
                    }
                    return resolve(success({
                      data: resData,
                      headers: res.headers,
                      retcode,
                      ...msgData,
                    }));
                  }
                  ctx.emit('cnfapi:res:reject', {
                    fnName: apiOpts.fnName,
                    req: {
                      headers: apiConfig.headers,
                      data: data.postData,
                      qs: data.qs,
                      params: data.getData,
                    },
                    ...(err || fail({
                      retcode: 500,
                      headers: res.headers,
                    })),
                  });
                  reject(err || fail({
                    retcode: 500,
                    headers: res.headers,
                  }));
                });
              }
              if (/^2\d/.test(+serverData.retcode)) {
                reqTime = 0;
                let { data } = serverData;
                /**
                 * [如果该接口有model配置，则根据model转换后输出]
                 */
                if (!isEmpty(apiConfig.model)) {
                  data = modelFn(apiConfig.model, data);
                }
                ctx.emit('cnfapi:res:resolve', {
                  fnName: apiOpts.fnName,
                  req: {
                    headers: apiConfig.headers,
                    data: data.postData,
                    qs: data.qs,
                    params: data.getData,
                  },
                  ...success({
                    data,
                    headers: res.headers,
                    retcode: serverData.retcode,
                  }),
                });
                return resolve(success({
                  data,
                  headers: res.headers,
                  retcode: serverData.retcode,
                }));
              }
              ctx.emit('cnfapi:res:reject', {
                fnName: apiOpts.fnName,
                req: {
                  headers: apiConfig.headers,
                  data: data.postData,
                  qs: data.qs,
                  params: data.getData,
                },
                ...fail({
                  retcode: serverData.retcode,
                  msg: serverData.msg,
                  headers: res.headers,
                }),
              });
              reject(fail({
                retcode: serverData.retcode,
                msg: serverData.msg,
                headers: res.headers,
              }));
              // retry(reqTime, retryTimes, interval, (isEnd) => {
              //     if (isEnd) {
              //         reqTime = 0;
              //         return reject(fail({
              //             retcode: serverData.retcode,
              //             msg: serverData.msg,
              //             headers: res.headers
              //         }));
              //     }
              //     request();
              // });
            }, (err) => {
              const errJSONMsg = JSON.stringify(err);
              if (errJSONMsg.toLowerCase().includes('abort')) {
                ctx.emit('cnfapi:res:reject', {
                  fnName: apiOpts.fnName,
                  req: {
                    headers: apiConfig.headers,
                    data: data.postData,
                    qs: data.qs,
                    params: data.getData,
                  },
                  ...fail({
                    retcode: retcode.OTHER,
                    msg: getType(err) === 'error' ? err.toString() : JSON.stringify(err),
                  }),
                });
                return reject(fail({
                  retcode: retcode.OTHER,
                  msg: getType(err) === 'error' ? err.toString() : JSON.stringify(err),
                }));
              }
              retry(reqTime, retryTimes, interval, (isEnd) => {
                if (isEnd) {
                  reqTime = 0;
                  ctx.emit('cnfapi:res:reject', {
                    req: {
                      headers: apiConfig.headers,
                      data: data.postData,
                      qs: data.qs,
                      params: data.getData,
                    },
                    fnName: apiOpts.fnName,
                    ...fail({
                      retcode: retcode.OTHER,
                      msg: getType(err) === 'error' ? err.toString() : JSON.stringify(err),
                    }),
                  });
                  return reject(fail({
                    retcode: retcode.OTHER,
                    msg: getType(err) === 'error' ? err.toString() : JSON.stringify(err),
                  }));
                }
                request(reqData);
              });
            }).catch((catchErr) => {
              reqTime = 0;
              ctx.emit('cnfapi:res:catch', {
                req: {
                  headers: apiConfig.headers,
                  data: data.postData,
                  qs: data.qs,
                  params: data.getData,
                },
                fnName: apiOpts.fnName,
                ...fail({
                  retcode: retcode.CATCH,
                  msg: getType(catchErr) === 'error' ? catchErr.toString() : JSON.stringify(catchErr),
                }),
              });
              reject(fail({
                retcode: retcode.CATCH,
                msg: getType(catchErr) === 'error' ? catchErr.toString() : JSON.stringify(catchErr),
              }));
            });
          });
        }

        function canRetryFn(interval, retryTimes) {
          if (isNaN(interval) || isNaN(retryTimes)) {
            return false;
          }
          interval = parseInt(interval);
          retryTimes = parseInt(retryTimes);
          if (interval <= 0 || retryTimes <= 0) {
            return false;
          }
          return true;
        }

        function retry(times, retryTimes, interval, cb = () => {
        }) {
          if (canRetryFn(interval, retryTimes) && reqTime <= retryTimes) {
            return setTimeout(() => {
              cb(times > retryTimes);
            }, interval);
          }
          cb(true);
        }
      }));
    },
  });
}

/**
 * success return data
 */
function success(res) {
  const msgData = {};
  if (res.msg) {
    msgData.msg = res.msg;
  }
  return {
    data: res.data,
    headers: res.headers,
    retcode: res.retcode,
    ...msgData,
  };
}

/**
 * fail return data
 */
function fail(res) {
  return {
    retcode: res.retcode || 'FE-5001',
    msg: res.msg || 'unknown',
    headers: res.headers || {},
  };
}

/**
 * serverData 2 FE data
 */
function modelFn(schema, data) {
  const feData = Array.isArray(data) ? [] : {};
  jsonSchema2api(feData, data, schema);
  return feData;
}

/**
 * 判断是否为空对象
 */
function isEmpty(obj) {
  return !obj || !Object.keys(obj).length;
}

/**
 * 获取数据类型
 */
function getType(val) {
  return Object.prototype.toString.call(val).slice(8, -1).toLowerCase();
}